### PR TITLE
TRUS-2288 Amend individual beneficiary

### PR DESCRIPTION
**Pull request by darrylmurphy**

### CODE REVIEW

File: app/extractors/IndividualBeneficiaryExtractor.scala [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        answers.set(DateOfBirthYesNoPage, false)
+    }
+  }
+
+  private def extractShareOfIncome(individualBeneficiary: IndividualBeneficiary,
+                                   answers: UserAnswers) : Try[UserAnswers] = {
+    individualBeneficiary.income match {
+      case Some(income) =>
+        answers.set(IncomeDiscretionYesNoPage, false)
+          // TODO come back and fix this toInt to handle at API reads
```
Comment (jackheslop96) [inline app/extractors/IndividualBeneficiaryExtractor.scala]: To do?

Comment (adamconder) [reply app/extractors/IndividualBeneficiaryExtractor.scala]: Well spotted! ;) I'll look at this now.

File: app/utils/mappers/AmendIndividualBeneficiaryMapper.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import pages.individualbeneficiary._
+import play.api.libs.functional.syntax._
+import play.api.libs.json.{JsError, JsSuccess, Reads}
+
+class AmendIndividualBeneficiaryMapper {
+
+  private val logger = LoggerFactory.getLogger("application." + this.getClass.getCanonicalName)
+
+  def apply(answers: UserAnswers): Option[IndividualBeneficiary] = {
+    val readFromUserAnswers: Reads[IndividualBeneficiary] =
```
Comment (jackheslop96) [inline app/utils/mappers/AmendIndividualBeneficiaryMapper.scala]: Could this be simplified to:

`val readFromUserAnswers: Reads[IndividualBeneficiary] =
      (
        NamePage.path.read[Name] and
          DateOfBirthPage.path.readNullable[LocalDate] and
          readIdentification and
          readAddress and
          VPE1FormYesNoPage.path.read[Boolean] and
          readIncome and
          IncomeDiscretionYesNoPage.path.read[Boolean] and
          StartDatePage.path.read[LocalDate] and
          Reads(_ => JsSuccess(true))
        ) .apply(IndividualBeneficiary.apply _ )`

File: app/utils/print/AmendIndividualBeneficiaryPrintHelper.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        bound.yesNoQuestion(DateOfBirthYesNoPage, "individualBeneficiary.dateOfBirthYesNo", controllers.individualbeneficiary.amend.routes.DateOfBirthYesNoController.onPageLoad().url),
+        bound.dateQuestion(DateOfBirthPage, "individualBeneficiary.dateOfBirth", controllers.individualbeneficiary.amend.routes.DateOfBirthController.onPageLoad().url),
+        bound.yesNoQuestion(IncomeDiscretionYesNoPage, "individualBeneficiary.incomeDiscretionYesNo", controllers.individualbeneficiary.amend.routes.IncomeDiscretionYesNoController.onPageLoad().url),
+        bound.intQuestion(IncomePercentagePage, "individualBeneficiary.incomePercentage", controllers.individualbeneficiary.amend.routes.IncomePercentageController.onPageLoad().url),
+        bound.yesNoQuestion(NationalInsuranceNumberYesNoPage, "individualBeneficiary.nationalInsuranceNumberYesNo", controllers.individualbeneficiary.amend.routes.NationalInsuranceNumberYesNoController.onPageLoad().url),
+        bound.ninoQuestion(NationalInsuranceNumberPage, "individualBeneficiary.nationalInsuranceNumber", controllers.individualbeneficiary.amend.routes.NationalInsuranceNumberYesNoController.onPageLoad().url),
```
Comment (jackheslop96) [inline app/utils/print/AmendIndividualBeneficiaryPrintHelper.scala]: Redirecting to wrong controller?

File: test/controllers/individualbeneficiary/remove/RemoveAddIndividualBeneficiaryControllerSpec.scala [hunk 4]
```diff
@@ -37,7 +37,7 @@ import views.html.individualbeneficiary.remove.RemoveIndexView
 
 import scala.concurrent.Future
 
-class RemoveIndividualBeneficiaryControllerSpec extends SpecBase with PropertyChecks with ScalaFutures {
+class RemoveAddIndividualBeneficiaryControllerSpec extends SpecBase with PropertyChecks with ScalaFutures {
```
Comment (jackheslop96) [inline test/controllers/individualbeneficiary/remove/RemoveAddIndividualBeneficiaryControllerSpec.scala]: Does this name make sense?

File: test/utils/print/AmendIndividualBeneficiaryPrintHelperSpec.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+          AnswerRow(label = Html(messages("individualBeneficiary.incomeDiscretionYesNo.checkYourAnswersLabel", name.displayName)), answer = Html("No"), changeUrl = controllers.individualbeneficiary.amend.routes.IncomeDiscretionYesNoController.onPageLoad().url),
+          AnswerRow(label = Html(messages("individualBeneficiary.incomePercentage.checkYourAnswersLabel", name.displayName)), answer = Html("50"), changeUrl = controllers.individualbeneficiary.amend.routes.IncomePercentageController.onPageLoad().url),
+          AnswerRow(label = Html(messages("individualBeneficiary.nationalInsuranceNumberYesNo.checkYourAnswersLabel", name.displayName)), answer = Html("Yes"), changeUrl = controllers.individualbeneficiary.amend.routes.NationalInsuranceNumberYesNoController.onPageLoad().url),
+          AnswerRow(label = Html(messages("individualBeneficiary.nationalInsuranceNumber.checkYourAnswersLabel", name.displayName)), answer = Html("AA 00 00 00 A"), changeUrl = controllers.individualbeneficiary.amend.routes.NationalInsuranceNumberYesNoController.onPageLoad().url),
```
Comment (jackheslop96) [inline test/utils/print/AmendIndividualBeneficiaryPrintHelperSpec.scala]: Wrong changeUrl

File: app/utils/mappers/AmendIndividualBeneficiaryMapper.scala (hunk 2 above)
Comment (adamconder) [reply app/utils/mappers/AmendIndividualBeneficiaryMapper.scala]: No as we need to default the provisional flag when we instantiate an instance of IndividualBeneficiary()

Comment (jackheslop96) [reply app/utils/mappers/AmendIndividualBeneficiaryMapper.scala]: The above code defaults it to true no?

**Pull request merged**